python/ql/src/experimental/Security/CWE-730/RegexDoS.ql

```diff
@@ -0,0 +1,86 @@
+/**
+ * @name Regular expression injection
+ * @description User input should not be used in regular expressions without first being escaped,
+ *              otherwise a malicious user may be able to inject an expression that could require
+ *              exponential time on certain inputs.
+ * @kind path-problem
+ * @problem.severity error
+ * @id python/regex-injection
+ * @tags security
+ *       external/cwe/cwe-730
+ *       external/cwe/cwe-400
+ */
+
+// determine precision above
+import python
+import semmle.python.dataflow.new.RemoteFlowSources
+import semmle.python.dataflow.new.DataFlow
+import semmle.python.dataflow.new.TaintTracking
+import semmle.python.dataflow.new.internal.TaintTrackingPublic
+import DataFlow::PathGraph
+
+// Should this be moved to a different structure? (For other queries to be able to use it)
+class ReMethods extends string {
+  ReMethods() {
+    this = "match" or
+    this = "fullmatch" or
+    this = "search" or
+    this = "split" or
+    this = "findall" or
+    this = "finditer"
+  }
+}
+
+class DirectRegex extends DataFlow::Node {
+  DirectRegex() {
+    exists(string reMethod, CallNode reCall |
+      reMethod instanceof ReMethods and
+      reCall = Value::named("re." + reMethod).getACall() and
+      this.asExpr() = reCall.getArg(0).getNode()
+    )
+  }
+}
+
+class CompiledRegex extends DataFlow::Node {
+  CompiledRegex() {
+    exists(CallNode patternCall, SsaVariable patternVar, CallNode reMethodCall |
+      patternCall = Value::named("re.compile").getACall() and
+      patternVar.getDefinition().getImmediateDominator() = patternCall and
+      patternVar.getAUse().getNode() = reMethodCall.getNode().getFunc().(Attribute).getObject() and
+      reMethodCall.getNode().getFunc().(Attribute).getName() instanceof ReMethods and
+      this.asExpr() = patternCall.getArg(0).getNode()
+    )
+  }
+}
```

Comment:
```suggestion
class CompiledRegex extends DataFlow::Node {
  CompiledRegex() {
    exists(DataFlow::CallCfgNode patternCall, AttrRead reMethod |
      patternCall = API::moduleImport("re").getMember("compile").getACall() and
      patternCall = reMethod.getObject().getALocalSource() and
      reMethod.getAttributeName() instanceof ReMethods and
      this = patternCall.getArg(0)
    )
  }
}
```
This will also find cases of the form `re.compile(regex).match` as well as cases with slightly more complicated flow. But primarily, it is using more principled technology :-)

Reply:
I'm learning a lot from these reviews, thank you so much!

Reply:
I'm having a hard time finding documentation on the `DataFlow::CallCfgNode` class. Can anyone point me in a direction to find that documentation? Thanks!

Reply:
Nvm, found it :)

https://github.com/github/codeql/blob/1ecee2da0d176abfe8db4096feabe8541c73c97b/docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst